fog/test-client/src/test_client.rs

```diff
@@ -826,3 +913,128 @@ impl Drop for ReceiveTxWorker {
+        self.inner.lock().unwrap().tx = Some(tx.clone());
+    }
+
+    /// Set the block count returned by tx_propose (immediately after it is
+    /// known)
+    pub fn set_tx_propose_block_count(&self, count: u64) {
+        self.inner.lock().unwrap().tx_propose_block_count = Some(count);
+    }
+
+    /// Set the index in which the tx appeared (immediately after it is known)
+    pub fn set_tx_appeared_block_index(&self, index: BlockIndex) {
+        self.inner.lock().unwrap().tx_appeared = Some(index);
+    }
+}
+
+impl core::fmt::Display for TxInfo {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        let guard = self.inner.lock().unwrap();
+        if let Some(proposed) = &guard.tx_propose_block_count {
+            write!(
+                f,
+                "Proposed at block index ~{}, ",
+                proposed.saturating_sub(1)
+            )?;
+        }
+        if let Some(appeared) = &guard.tx_appeared {
+            write!(f, "Appeared in block index {}, ", appeared)?;
+        }
+        if let Some(tx) = &guard.tx {
+            write!(
+                f,
+                "TxOut public keys: [{}]",
+                HexList(tx.prefix.outputs.iter().map(|x| x.public_key.as_bytes()))
+            )?;
+        }
+        Ok(())
+    }
+}
+
+/// Object which manages the LAST_POLLING_SUCCESSFUL gauge
+///
+/// * If a failure is observed, we are unhealthy (immediately)
+/// * If no failure is observed for a long enough time, we are healthy again
+///
+/// The amount of time is the "healing_time" and it is expected to be set
+/// to the number of clients, so that when we go around once in round-robin
+/// fashion and all clients are successful, we are considered healed, and not
+/// before that.
+#[derive(Default)]
+pub struct HealthTracker {
+    // Set to i for the duration of the i'th transfer
+    counter: AtomicUsize,
+    // Whether we have observed any failure
+    have_failure: AtomicBool,
+    // The counter value during the most recent failure
+    last_failure: AtomicUsize,
+    // Healing time: How many successful transfers needed to forget a failure
+    healing_time: usize,
+}
+
+impl HealthTracker {
+    /// Make a new healthy tracker.
+    /// Sets LAST_POLLING_SUCCESSFUL to true initially.
+    ///
+    /// Takes "healing time" which is the number of successful transfers before
+    /// we consider ourselves healthy again
+    pub fn new(healing_time: usize) -> Self {
+        counters::LAST_POLLING_SUCCESSFUL.set(1);
+        Self {
+            healing_time,
+            ..Default::default()
+        }
+    }
+
+    /// Set the counter value, and maybe update healthy status
+    pub fn set_counter(&self, counter: usize) {
+        self.counter.store(counter, Ordering::SeqCst);
+        // If:
+        // * there is a failure
+        // * the failure happened at least healing_time ago
+        // then set ourselves healthy again
+        if self.have_failure.load(Ordering::SeqCst)
```

Comment:
This is probably a dumb question, but this implies that counter can only be set to a larger value than the current value right? Like it should never be the case that we set counter to before the healing time and return to an unhealed state?

Reply:
yeah its an assumption of the code -- i should probly document it